orangecontrib/prototypes/widgets/contingency_table.py

```diff
@@ -0,0 +1,365 @@
+import unicodedata
+from math import isnan, isinf, sqrt, pi
+
+from AnyQt.QtCore import Qt, QItemSelection, QItemSelectionModel
+from AnyQt.QtGui import QStandardItem, QColor, QFont, QBrush, QPainter
+from AnyQt.QtWidgets import QTableView, QSizePolicy, QHeaderView, QStyledItemDelegate
+from Orange.widgets import gui
+
+BorderRole = next(gui.OrangeUserRole)
+BorderColorRole = next(gui.OrangeUserRole)
+
+CircleArea = next(gui.OrangeUserRole)
+
+
+class BorderedItemDelegate(QStyledItemDelegate):
+    """Item delegate that paints border at the specified sides
+
+    Data for `BorderRole` is a string containing letters t, r, b and/or l,
+    which defines the sides at which the border is drawn.
+
+    Role `BorderColorRole` sets the color for the cell. If not color is given,
+    `self.color` is used as default.
+
+    Args:
+        color (QColor): default color (default default is black)
+    """
+    def __init__(self, color=Qt.black):
+        super().__init__()
+        self.color = color
+
+    def paint(self, painter, option, index):
+        """Overloads `paint` to draw borders"""
+        QStyledItemDelegate.paint(self, painter, option, index)
+        borders = index.data(BorderRole)
+        if borders:
+            color = index.data(BorderColorRole) or self.color
+            painter.save()
+            painter.setPen(color)
+            rect = option.rect
+            for side, p1, p2 in (("t", rect.topLeft(), rect.topRight()),
+                                 ("r", rect.topRight(), rect.bottomRight()),
+                                 ("b", rect.bottomLeft(), rect.bottomRight()),
+                                 ("l", rect.topLeft(), rect.bottomLeft())):
+                if side in borders:
+                    painter.drawLine(p1, p2)
+            painter.restore()
+
+
+class CircleItemDelegate(BorderedItemDelegate, gui.VerticalItemDelegate):
+    def sizeHint(self, option, index):
+        if index.column() == 0 or index.row() == 1:
```

Comment:
Could you restructure this to first handle header rows/columns and use QStyleItemDelegates implementation for everything else? As far as I can tell, col 0 and row 1 are vertical, column 1 and row 0 are bordered, everything else is Styled.